federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -0,0 +1,431 @@
+func (d *ServiceDNSController) getUplevelCName(dnsName, dnsNamePrefix string) (uplevelCname string) {
+	// split the dnsName with . as separator to get slice of strings
+	dnsNameSlice := strings.Split(dnsName, ".")
+	// find the index of the string which needs to be removed, so that we get uplevel CNAME
+	indexToDelete := len(strings.Split(dnsNamePrefix, "."))
+	// remove the zone/region from dnsName
+	dnsNameSlice = append(dnsNameSlice[:indexToDelete], dnsNameSlice[indexToDelete+1:]...)
+
+	uplevelCname = strings.Join(dnsNameSlice, ".")
+
+	// uplevelCname should have serviceDNSSuffix, otherwise the input dnsName was a global level name, so return empty string
+	if !strings.HasSuffix(uplevelCname, d.serviceDnsSuffix) {
+		uplevelCname = ""
+	}
+	return uplevelCname
+}
+
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets are consistent with endpoints.
+ */
+func (d *ServiceDNSController) ensureDnsRrsets(dnsZone dnsprovider.Zone, endpoints map[string][]string, dnsNamePrefix string) error {
+	rrsets, supported := dnsZone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("DNS provider does not support the ResourceRecordSets interface")
+	}
+
+	changeSet := rrsets.StartChangeset()
+
+	for dnsName, endpoints := range endpoints {
+		rrset, err := rrsets.Get(dnsName)
+		if err != nil {
+			return err
+		}
+
+		if len(endpoints) < 1 {
+			// There are no healthy endpoints for the dnsName, So need to add an appropriate CNAME record.
+			uplevelCname := d.getUplevelCName(dnsName, dnsNamePrefix)
+			glog.V(4).Infof("There are no healthy endpoint addresses at level %q, so CNAME to %q, if provided", dnsName, uplevelCname)
+			if uplevelCname != "" {
+				// dnsName is for zone or region level
+				// should ensure a CNAME pointing to one level above
+				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
+				if rrset == nil {
+					// rrset does not exist, so need to create new CNAME Record
+					glog.V(4).Infof("Creating CNAME to %q for %q", uplevelCname, dnsName)
+					changeSet.Add(newRrset) // the rrset already exists, ensure that it is right.
+
+				} else {
+					if !dnsprovider.ResourceRecordSetsEquivalent(rrset, newRrset) {
+						glog.V(4).Infof("Have recordset %v. Need recordset %v", rrset, newRrset)
+						changeSet.Remove(rrset).Add(newRrset)
+					} else {
+						glog.V(4).Infof("Existing recordset %v is equivalent to needed recordset %v, our work is done here.", rrset, newRrset)
+					}
+				}
+			} else {
+				// dnsName is for global level
+				if rrset == nil {
+					glog.V(4).Infof("We want no record for %q, and we have no record, so we're all good.", dnsName)
+				} else {
+					// the rrset exists, so should remove it as no healthy endpoints exist now.
+					glog.V(4).Infof("Successfully removed existing recordset %v", rrset)
```

Comment:
nit: This log message is not quite right.  You haven't successfully removed it yet.  Rather something like "Appending removal of recordset %v to change set."